src/data/rat/cast.lean

```diff
@@ -277,3 +277,52 @@ end
 
 instance rat.subsingleton_ring_hom {R : Type*} [semiring R] : subsingleton (ℚ →+* R) :=
 ⟨ring_hom.ext_rat⟩
+
+/-- Integer values of a morphism `φ` completely determine `φ`. -/
+theorem ext_hom_int {α} [group_with_zero α] {φ₁ φ₂ : monoid_with_zero_hom ℚ α}
+  (same_on_int : φ₁.comp (int.cast_ring_hom ℚ).to_monoid_with_zero_hom =
+    φ₂.comp (int.cast_ring_hom ℚ).to_monoid_with_zero_hom) : φ₁ = φ₂ :=
+begin
+  have same_on_int' : ∀ k : ℤ, φ₁ k = φ₂ k,
+  from monoid_with_zero_hom.congr_fun same_on_int,
+  ext x,
+  rw ← @rat.num_denom x,
+  rw rat.mk_eq_div,
+  repeat { rw div_eq_mul_inv, },
+  rw ← coe_coe,
+  rw monoid_with_zero_hom.map_mul φ₁,
+  rw monoid_with_zero_hom.map_inv' φ₁ x.denom,
+  rw monoid_with_zero_hom.map_mul φ₂,
+  rw monoid_with_zero_hom.map_inv' φ₂ x.denom,
+  rw same_on_int' x.num,
+  have := same_on_int' x.denom, rw ← coe_coe at this,
+  rw this,
+end
+
+theorem ext_hom_pnat {α} [group_with_zero α] {φ₁ φ₂ : monoid_with_zero_hom ℤ α}
+  (same_on_neg_one : φ₁ (-1) = φ₂ (-1)) (same_on_pnat : ∀ {n: ℕ}, 0 < n → φ₁ n = φ₂ n) : φ₁ = φ₂ :=
```

Comment:
This probably belongs in `data/int/cast` or similar, since this lemma doesn't mention rational numbers at all